crates/bevy_math/src/common_traits.rs

```diff
@@ -161,3 +161,133 @@ impl NormedVectorSpace for f32 {
         self * self
     }
 }
+
+/// A type that can be intermediately interpolated between two given values
```

Comment:
I'm slightly concerned about the fact we're adding to a foundation library a trait which is very vague, and whose implementation is relied upon by another library (animation) so is making implicit assumptions in its implementation for math types (directions and rotations use `slerp`, other types use `lerp`). I understand the drive to uniformize and remove duplicates code, but what happens when someone wants to use that trait in another library? There's almost zero guarantee on the produced value other than some vague "interpolation" mention, so if I were a crate author I'd avoid that trait because I can't reason about it. While it's located in the animation library at least there's an assumption it's used for animation, and therefore that the interpolation is linear. But here, nothing prevents me from implementing this trait as some weird random value provided I satisfy the constraints at t=0 and t=1.

Reply:
To me, using `lerp`/`slerp` for linear and rotational types has anything to do with animation *per se* — those are the modes of interpolation that are mathematically natural; my expectation is that if `bevy_animation` wanted something else specific to their use-cases, they would newtype them to customize the interpolation. In my opinion, the only concession here is for `bool`, but perhaps the better way forward would be to just force `bevy_animation` to newtype it. 

As to the matter of vagueness: this is at least partially by design; for some additional context, this trait's real reason for existence is to allow for quite varied notions of interpolation, especially in light of the future [Curve RFC](https://github.com/bevyengine/rfcs/pull/80), which presents a general framework wherein this trait allows curves to be consistently defined by discrete data. That was designed primarily with the needs of future work in `bevy_math` in mind (i.e. curve geometry, RMFs, reparametrizing motion, and so on), but the idea is that `bevy_animation` would also use it for curves in animation. 

I'm not sure exactly what the concern about API consumers is; I would expect that calling `interpolate` on vectors or directions should produce an intuitive notion of interpolation between them, and it does. If third-party libraries want to implement weird implementations of interpolation, there is nothing anyone can do to stop them regardless, but that's true of all public trait interfaces.

On some level, I do think this is almost lower-level than `bevy_math` (i.e. `bevy_math` is kind of more of an API consumer than a producer), but I don't think there's really any better place for it right now.

Reply:
> if bevy_animation wanted something else specific to their use-cases, they would newtype them to customize the interpolation

Probably impossible, Bevy Animation works on the types it's asked to work on by the user, not the types it wants to. 

> has anything to do with animation per se

Maybe, but animation relies on linear interpolation. If we changed it, everything would break. And this trait doesn't provide any assurance about linearity.

> I'm not sure exactly what the concern about API consumers is; I would expect that calling interpolate on vectors or directions should produce an intuitive notion of interpolation between them, and it does. 

This explanation highlights very precisely what I'm worried about, the only definition here is "intuitive" and that's subjective. There's millions of values between `0.` and `1.`, all of them can be defined as an interpretation of `0.` and `1.` via some function. If we don't specify the function, and just say "intuitive", then the trait is unusable because what's intuitive depends on the usage context, and it's not intrinsic such that we can have a single implementation.

> this trait's real reason for existence is to allow for quite varied notions of interpolation

I don't understand how a trait with a single important per type can allow varied notions. The traditional way of doing non linear interpolation, like Bevy Animation, is to likely interpolate values using a non linear parameter `t`. But then your other comment states that it's linear, which prevents this pattern.

Reply:
> Probably impossible, Bevy Animation works on the types it's asked to work on by the user, not the types it wants to.

Actually quite easy, considering that Curves are mappable: the underlying data that is being interpolated does not need to have the same type as its output.

> Maybe, but animation relies on linear interpolation. If we changed it, everything would break. And this trait doesn't provide any assurance about linearity.

No, but it makes very strong demands on interpolation nonetheless. 

> This explanation highlights very precisely what I'm worried about, the only definition here is "intuitive" and that's subjective. There's millions of values between `0.` and `1.`, all of them can be defined as an interpretation of `0.` and `1.` via some function. If we don't specify the function, and just say "intuitive", then the trait is unusable because what's intuitive depends on the usage context, and it's not intrinsic such that we can have a single implementation.

"Some function" is not likely to satisfy the requirements of the trait implementation, which are quite strict.